test/e2e/upgrades/apps/replicasets.go

```diff
@@ -0,0 +1,145 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package upgrades
+
+import (
+	"fmt"
+	"reflect"
+	"time"
+
+	extensions "k8s.io/api/extensions/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/kubernetes/pkg/util/version"
+	"k8s.io/kubernetes/test/e2e/framework"
+	"k8s.io/kubernetes/test/e2e/upgrades"
+
+	. "github.com/onsi/ginkgo"
+	imageutils "k8s.io/kubernetes/test/utils/image"
+)
+
+const (
+	interval = 2 * time.Second
+	timeout  = 1 * time.Minute
+	rsName   = "replicaset-hash-test"
+)
+
+// TODO: Test that the replicaset stays available during master (and maybe
+// node and cluster upgrades).
+
+// ReplicaSetUpgradeTest tests that a replicaset survives upgrade.
+type ReplicaSetUpgradeTest struct {
+	oldRSStatus extensions.ReplicaSetStatus
+	newRSStatus extensions.ReplicaSetStatus
+}
+
+func (ReplicaSetUpgradeTest) Name() string { return "[sig-apps] replicaset-upgrade" }
+
+func (ReplicaSetUpgradeTest) Skip(upgCtx upgrades.UpgradeContext) bool {
+	// As of 1.8, the client code we call into no longer supports talking to a server <1.7. (see #47685)
+	minVersion := version.MustParseSemantic("v1.7.0")
+
+	for _, vCtx := range upgCtx.Versions {
+		if vCtx.Version.LessThan(minVersion) {
+			return true
+		}
+	}
+	return false
+}
+
+var _ upgrades.Skippable = ReplicaSetUpgradeTest{}
+
+// Setup creates a replicaset and makes sure it has a new and an old replicaset running.
+// This calls in to client code and should not be expected to work against a cluster more than one minor version away from the current version.
+func (r *ReplicaSetUpgradeTest) Setup(f *framework.Framework) {
+	c := f.ClientSet
+	ns := f.Namespace.Name
+	nginxImage := imageutils.GetE2EImage(imageutils.NginxSlim)
+
+	By(fmt.Sprintf("Creating a replicaset %q in namespace %q", rsName, ns))
+	replicaSet := framework.NewReplicaSet(rsName, ns, int32(1), map[string]string{"test": "upgrade"}, "nginx", nginxImage)
+	rs, err := c.Extensions().ReplicaSets(ns).Create(replicaSet)
+	framework.ExpectNoError(err)
+
+	By(fmt.Sprintf("Waiting replicaset %q's container image to be updated", rsName))
+	err = framework.WaitForReplicaSetImage(c, ns, rsName, nginxImage, interval, timeout)
+	framework.ExpectNoError(err)
+
+	By(fmt.Sprintf("Waiting for replicaset %q to have all of its replicas ready.", rsName))
+	framework.ExpectNoError(framework.WaitForReadyReplicaSet(c, ns, rsName))
+
+	// Store the old RS status.
+	r.oldRSStatus = rs.Status
+	oldUID := rs.UID
+
+	// Trigger a new rollout so that we have some history.
```

Comment:
ReplicaSet doesn't support rollout.

Reply:
The comment is wrong. It should be `Change pod template so that it can be shown that the upgrade reflects newest change`. Correct me if I am wrong here.

Reply:
What I mean is, because RS doesn't support rollout, nothing will happen to existing Pods when you change the Pod template in the RS spec. So there is no equivalent to the steps in Deployment tests where it checks that rollout works. You can skip anything rollout-related.

Reply:
In fact, I didn't know (and am surprised) that nothing happens to existing pods when we change pod template in RS spec. I will remove all codes and comments related to rollout.